src/LegalizeVectors.cpp

```diff
@@ -0,0 +1,630 @@
+                assignments.push_back(Store::make(
+                    op->name, std::move(rhs), std::move(index),
+                    op->param, std::move(predictate), op->alignment + lane_start));
+            }
+            Stmt result = Block::make(assignments);
+            debug(3) << "Legalized store " << Stmt(op) << " => " << result << "\n";
+            return result;
+        }
+        return IRMutator::visit(op);
+    }
+
+    Expr visit(const Shuffle *op) override {
+        if (max_lanes == 0) {
+            return IRMutator::visit(op);
+        }
+        internal_assert(op->type.lanes() <= max_lanes) << Expr(op);
+        bool requires_mutation = false;
+        for (size_t i = 0; i < op->vectors.size(); ++i) {
+            if (op->vectors[i].type().lanes() > max_lanes) {
+                requires_mutation = true;
+                break;
+            }
+        }
+
+        if (requires_mutation) {
+            debug(4) << "Legalizing Shuffle " << Expr(op) << "\n";
+            // We are dealing with a shuffle of an exceeding-lane-count vector argument.
+            // We can assume the variable here has extracted lane variables in surrounding Lets.
+            // So let's hope it's a simple case, and we can legalize.
+
+            vector<Expr> new_vectors;
+            vector<pair<int, int>> vector_and_lane_indices = op->vector_and_lane_indices();
+            for (int i = 0; i < int(op->vectors.size()); ++i) {
+                const Expr &vec = op->vectors[i];
+                if (vec.type().lanes() > max_lanes) {
+                    debug(4) << "  Arg " << i << ": " << vec << "\n";
+                    int num_vecs = (vec.type().lanes() + max_lanes - 1) / max_lanes;
+                    for (int i = 0; i < num_vecs; i++) {
+                        int lane_start = i * max_lanes;
+                        int lane_count_for_vec = std::min(vec.type().lanes() - lane_start, max_lanes);
+                        new_vectors.push_back(ExtractLanes(lane_start, lane_count_for_vec, max_lanes).mutate(vec));
+                    }
+                } else {
+                    new_vectors.push_back(IRMutator::mutate(vec));
+                }
+            }
+            Expr result = Shuffle::make(new_vectors, op->indices);
+            result = simplify_shuffle(result.as<Shuffle>());
+            debug(3) << "Legalized " << Expr(op) << " => " << result << "\n";
+            return result;
+        }
+        return IRMutator::visit(op);
+    }
+
+    Expr visit(const VectorReduce *op) override {
+        if (max_lanes == 0) {
+            return IRMutator::visit(op);
+        }
+        const Expr &arg = op->value;
+        if (arg.type().lanes() > max_lanes) {
+            int vecs_per_reduction = op->value.type().lanes() / op->type.lanes();
+            if (vecs_per_reduction % max_lanes == 0) {
+                // This should be possible too. TODO
```

Comment:
I'm a bit confused by what this case represents. Is this TODO something that needs to be handled?

Reply:
So confusing it's just wrong. I deleted it.